### PR TITLE
IndexedStore

### PR DESCRIPTION
**Pull request by wemrysi**
Introduces the `IndexedStore` abstraction and includes an implementation backed by `mimir`.

### CODE REVIEW

File: impl/src/main/scala/quasar/impl/storage/IndexedStore.scala [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+          def insert(i: I, v: V): B[Unit] =
+            f(fa.insert(i, v))
+
+          def delete(i: I): B[Unit] =
+            f(fa.delete(i))
+        }
+    }
+
+  implicit def indexInvariantFunctor[F[_]: Functor, V]: InvariantFunctor[IndexedStore[F, ?, V]] =
```
Comment (djspiewak) [inline impl/src/main/scala/quasar/impl/storage/IndexedStore.scala]: Are you sure you want this? It's never going to be selected, given that the parameter is not right-most. It's also technically incoherent with the right-associated one below.

**Comment by wemrysi** [comment]
[ch519]

File: mimir/src/main/scala/quasar/mimir/storage/MimirIndexedStore.scala [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      rpath <- Stream.emits(rpaths).covary[F]
+
+      rfile <- Stream.emit(Path.refineType(rpath).toOption).unNone.covary[F]
+
+      s = Path.fileName(rfile).value
+
+      v <- Stream.eval(lookup(s)).unNone
+    } yield (s, v)
+
+  def lookup(s: String): F[Option[RValue]] = {
```
Comment (djspiewak) [inline mimir/src/main/scala/quasar/mimir/storage/MimirIndexedStore.scala]: Shouldn't we at least newtype the key?

File: mimir/src/main/scala/quasar/mimir/storage/MimirIndexedStore.scala [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+  import precog.trans.constants._
+
+  val entries: Stream[F, (String, RValue)] =
+    for {
+      rpaths <- Stream.eval(precog.vfs.ls(tablesPrefix).to[F])
+
+      rpath <- Stream.emits(rpaths).covary[F]
+
+      rfile <- Stream.emit(Path.refineType(rpath).toOption).unNone.covary[F]
```
Comment (djspiewak) [inline mimir/src/main/scala/quasar/mimir/storage/MimirIndexedStore.scala]: I'm in favor of newlines, but, subjectively… I would collapse these three lines together.

File: mimir/src/test/scala/quasar/mimir/storage/MimirIndexedStoreSpec.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    } yield result
+
+    init.unsafeRunSync
+  }
+
+  val emptyStore =
+    IO(Random.nextInt(100000)) map { i =>
+      MimirIndexedStore(
+        P.unsafeValue,
+        rootDir </> dir("store") </> dir(i.toString))
```
Comment (djspiewak) [inline mimir/src/test/scala/quasar/mimir/storage/MimirIndexedStoreSpec.scala]: I would propose that we standardize on using `.store` and similar for "meta" storage within the mimir vfs, just to help us keep it straight. The counter-argument here is that we don't really put things raw into mimir anymore, so maybe it's not necessary?

Comment (wemrysi) [reply mimir/src/test/scala/quasar/mimir/storage/MimirIndexedStoreSpec.scala]: Doesn't matter to me, the production locations are specified in constants somewhere very close to `Main` and they're not exposed to any clients, so the specifics don't matter terribly, but :+1: to standardizing on something.

File: foundation/src/test/scala/quasar/EffectfulQSpec.scala [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+abstract class EffectfulQSpec[F[_]: Effect](implicit ec: ExecutionContext) extends Qspec {
+  /** Provides syntax for defining effectful examples:
+    *
+    * "some example name" >>* {
+    *   <example body returning F[A]>
+    * }
+    */
+  implicit class RunExample(s: String) {
+    def >>*[A: AsResult](fa: => F[A]): Fragment = {
```
Comment (alissapajer) [inline foundation/src/test/scala/quasar/EffectfulQSpec.scala]: :+1:

File: impl/src/main/scala/quasar/impl/storage/IndexedStore.scala (hunk 1 above)
Comment (wemrysi) [reply impl/src/main/scala/quasar/impl/storage/IndexedStore.scala]: Yeah, I know it won't be selected normally, but good point about the incoherence. I'll just make it a function.

File: mimir/src/test/scala/quasar/mimir/storage/MimirIndexedStoreSpec.scala (hunk 4 above)
Comment (djspiewak) [reply mimir/src/test/scala/quasar/mimir/storage/MimirIndexedStoreSpec.scala]: On further thought, I'm fine with just `store`

**Pull request merged**